clippy_lints/src/path_buf_push_overwrite.rs

```diff
@@ -0,0 +1,65 @@
+use crate::utils::{match_type, paths, span_help_and_lint, walk_ptrs_ty};
+use if_chain::if_chain;
+use rustc::hir::*;
+use rustc::lint::{LateContext, LateLintPass, LintArray, LintPass};
+use rustc::{declare_tool_lint, lint_array};
+use std::path::Path;
+use syntax::ast::LitKind;
+
+declare_clippy_lint! {
+    /// **What it does:*** Checks for `push` calls on `PathBuf` that can cause overwrites.
+    ///
+    /// **Why is this bad?** Calling `push` with a root path at the start can overwrite the
+    /// previous defined path.
+    ///
+    /// **Known problems:** None.
+    ///
+    /// **Example:**
+    /// ```ignore
+    /// let mut x = PathBuf::from("/foo");
+    /// x.push("/bar");
+    /// println!("x: {}", x.display()); // "/bar
```

Comment:
You can replace this `println!` with an `assert!`, so you don't have to write a comment.
```suggestion
    /// assert_eq!(x, PathBuf::from("/bar")); // "/bar
```

The println below can be replaced also